### PR TITLE
Feature/expose fpe in interfaces

### PR DESCRIPTION
**Pull request by Manuthor**
Expose FFI, Webassembly and Pyo3 interfaces for Format-Preserving-Encryption (FPE):
- [x] add integer and big integers encryption (as string with radix and digits)
- [x] add string encryption according to given alphabet
  - "numeric": 0123456789
  - "hexa_decimal": 0123456789abcdef
  - "alpha_lower": abcdefghijklmnopqrstuvwxyz
  - "alpha_upper": ABCDEFGHIJKLMNOPQRSTUVWXYZ
  - "alpha": abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ
  - "alpha_numeric": 0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ
  - "utf": creates an Alphabet with the first 63489 (~2^16) Unicode characters
  - "chinese": creates an Alphabet with the Chinese characters
  - "latin1sup": creates an Alphabet with the latin-1 and latin1-supplement characters (supports French)
  - "latin1sup_alphanum": creates an Alphabet with the latin-1 and latin1-supplement characters but without the non alphanumeric characters (supports French)
- [x] add float encryption

### CODE REVIEW

File: crates/fpe/python/cloudproof_fpe/__init__.pyi [hunk 1]
```diff
@@ -0,0 +1,133 @@
+from typing import List, Optional, Tuple, Union
```
Comment (ackRow) [inline crates/fpe/python/cloudproof_fpe/__init__.pyi:1]:
```suggestion

```
I think these imports are not used

File: crates/fpe/python/cloudproof_fpe/__init__.pyi [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        """ """
+    def encrypt_big(
+        self,
+        key: bytes,
+        tweak: bytes,
+        plaintext: str,
+    ) -> str:
+        """ """
+    def decrypt_big(self, key: bytes, tweak: bytes, ciphertext: str) -> str:
+        """ """
```
Comment (ackRow) [inline crates/fpe/python/cloudproof_fpe/__init__.pyi:95]: Empty docstring

**Review by ackRow — COMMENTED** [review]
Are the tests in `fpe/python/tests/fpe_test.py` checked in the CI ?

Maybe add `test_python_interface fpe` in https://github.com/Cosmian/cloudproof_rust/blob/main/scripts/test_python.sh#L35

File: crates/fpe/src/ffi/alphabet.rs [hunk 3]
```diff
@@ -0,0 +1,163 @@
+use std::ffi::{c_char, c_int, c_uchar};
+
+use cosmian_ffi_utils::{ffi_read_bytes, ffi_read_string, ffi_unwrap, ffi_write_bytes};
+
+use crate::get_alphabet;
+
+#[allow(clippy::too_many_arguments)]
+pub(crate) unsafe fn fpe_alphabet_internal(
```
Comment (ccorsin) [inline crates/fpe/src/ffi/alphabet.rs:8]: call it "fpe" to stay consistent with Float and Integer ffi's files ?

File: crates/fpe/src/wasm_bindgen/alphabet.rs [hunk 4]
```diff
@@ -0,0 +1,47 @@
+use wasm_bindgen::{prelude::wasm_bindgen, JsValue};
+
+use crate::get_alphabet;
+
+fn fpe_process(
```
Comment (ccorsin) [inline crates/fpe/src/wasm_bindgen/alphabet.rs:5]: same - call it "fpe" ?

File: crates/fpe/src/wasm_bindgen/float.rs [hunk 5]
```diff
@@ -0,0 +1,48 @@
+use wasm_bindgen::{prelude::wasm_bindgen, JsValue};
+
+use crate::core::{Float, KEY_LENGTH};
+
+fn encrypt_decrypt(
```
Comment (ccorsin) [inline crates/fpe/src/wasm_bindgen/float.rs:5]: same - not constistent

File: crates/fpe/src/wasm_bindgen/integer.rs [hunk 6]
```diff
@@ -0,0 +1,127 @@
+use num_bigint::BigUint;
+use num_traits::Num;
+use wasm_bindgen::{prelude::wasm_bindgen, JsValue};
+
+use crate::core::{Integer, KEY_LENGTH};
+
+fn encrypt_decrypt(
```
Comment (ccorsin) [inline crates/fpe/src/wasm_bindgen/integer.rs:7]: same - not consistent

File: crates/fpe/python/cloudproof_fpe/__init__.pyi [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+    def __init__(self, alphabet_id: str):
+        """
+        Initialize the FormatPreservingEncryption instance with a given
+        alphabet type.
+
+        :param alphabet_type: A string indicating the type of alphabet to be used
+        for the encryption. Currently, only the following options are supported:
+        'alpha', 'alpha_lower', 'alpha_numeric', 'alpha_numeric_sym', 'latin1',
+        'latin1_sup', 'numeric', 'hex', and 'base64'.
```
Comment (ccorsin) [inline crates/fpe/python/cloudproof_fpe/__init__.pyi:17]: not consistent with   Numeric = "numeric",
  Hexadecimal = "hexa_decimal",
  AlphaLower = "alpha_lower",
  AlphaUpper = "alpha_upper",
  Alpha = "alpha",
  AlphaNumeric = "alpha_numeric",
  Utf = "utf",
  Chinese = "chinese",
  Latin1Sup = "latin1sup",
  Latin1SupAlphanum = "latin1sup_alphanum",

Comment (Manuthor) [reply crates/fpe/python/cloudproof_fpe/__init__.pyi:17]: thanks

File: crates/fpe/src/wasm_bindgen/integer.rs (hunk 6 above)
Comment (Manuthor) [reply crates/fpe/src/wasm_bindgen/integer.rs:7]: done

**Comment by Manuthor** [comment]
> Are the tests in `fpe/python/tests/fpe_test.py` checked in the CI ?
> 
> Maybe add `test_python_interface fpe` in https://github.com/Cosmian/cloudproof_rust/blob/main/scripts/test_python.sh#L35

indeed, this script `test_python` wasn't tested in CI (but python code was tested from `reusable_workflows`). It is now.
Thanks

**Review by ackRow — APPROVED** [review]

File: crates/fpe/src/ffi/float.rs [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    tweak_ptr: *const c_char,
+    tweak_len: c_int,
+    encrypt_flag: bool,
+) -> c_int {
+    let key_bytes = ffi_read_bytes!("key", key_ptr, key_len);
+    let tweak_bytes = ffi_read_bytes!("tweak", tweak_ptr, tweak_len);
+
+    // Copy the contents of the slice into the 32-array
+    let mut key: [u8; 32] = [0; 32];
+    key.copy_from_slice(key_bytes);
```
Comment (ThibaudDauce) [inline crates/fpe/src/ffi/float.rs:21]: I think there is a `TryFrom` implementation to do that… (and throw an error if the size is not correct)

File: crates/fpe/src/ffi/integer.rs [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+    tweak_ptr: *const c_char,
+    tweak_len: c_int,
+    encrypt_flag: bool,
+) -> c_int {
+    let key_bytes = ffi_read_bytes!("key", key_ptr, key_len);
+    let tweak_bytes = ffi_read_bytes!("tweak", tweak_ptr, tweak_len);
+
+    // Copy the contents of the slice into the 32-array
+    let mut key: [u8; 32] = [0; 32];
+    key.copy_from_slice(key_bytes);
```
Comment (ThibaudDauce) [inline crates/fpe/src/ffi/integer.rs:25]: Use `TryFrom` implementation

File: crates/fpe/src/ffi/integer.rs [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+    );
+    let input_biguint = ffi_unwrap!(
+        BigUint::from_str_radix(&input_str, radix),
+        "failed to convert
+input to BigUint"
+    );
+
+    // Copy the contents of the slice into the 32-array
+    let mut key: [u8; 32] = [0; 32];
+    key.copy_from_slice(key_bytes);
```
Comment (ThibaudDauce) [inline crates/fpe/src/ffi/integer.rs:155]: Use `TryFrom` implementation

File: crates/fpe/src/wasm_bindgen/alphabet.rs [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+    alphabet_id: &str,
+    key: Vec<u8>,
+    tweak: Vec<u8>,
+    additional_chars: &str,
+    encrypt_flag: bool,
+) -> Result<String, JsValue> {
+    let mut alphabet =
+        get_alphabet(alphabet_id).map_err(|e| JsValue::from_str(&format!("{e:?}")))?;
+
+    // let mut alphabet = alphabet.clone();
```
Comment (ThibaudDauce) [inline crates/fpe/src/wasm_bindgen/alphabet.rs:16]: Remove this comment?

File: crates/fpe/src/wasm_bindgen/float.rs [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+fn fpe(input: f64, key: Vec<u8>, tweak: Vec<u8>, encrypt_flag: bool) -> Result<f64, JsValue> {
+    if key.len() != KEY_LENGTH {
+        return Err(JsValue::from_str(&format!(
+            "FPE Float error: key length incorrect: {}, expected {}",
+            key.len(),
+            KEY_LENGTH
+        )));
+    }
+    let mut k: [u8; 32] = [0; 32];
+    k.copy_from_slice(&key);
```
Comment (ThibaudDauce) [inline crates/fpe/src/wasm_bindgen/float.rs:14]: Use `TryFrom` implementation

File: crates/fpe/src/wasm_bindgen/integer.rs [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+        return Err(JsValue::from_str(&format!(
+            "FPE Big Integer error: key length incorrect: {}, expected {}",
+            key.len(),
+            KEY_LENGTH
+        )));
+    }
+
+    // Copy the key bytes into a 32-byte array
+    let mut k: [u8; 32] = [0; 32];
+    k.copy_from_slice(&key);
```
Comment (ThibaudDauce) [inline crates/fpe/src/wasm_bindgen/integer.rs:26]: Use `TryFrom` implementation

Comment (ThibaudDauce) [reply crates/fpe/src/wasm_bindgen/integer.rs:26]: You can also remove the condition above with the `try_into`

File: crates/fpe/src/wasm_bindgen/float.rs (hunk 12 above)
Comment (ThibaudDauce) [reply crates/fpe/src/wasm_bindgen/float.rs:14]: You can remove the check above with `try_into()`

File: crates/fpe/src/ffi/float.rs (hunk 8 above)
Comment (Manuthor) [reply crates/fpe/src/ffi/float.rs]: thanks

File: crates/fpe/src/wasm_bindgen/alphabet.rs (hunk 11 above)
Comment (Manuthor) [reply crates/fpe/src/wasm_bindgen/alphabet.rs]: done

File: crates/fpe/src/ffi/integer.rs (hunk 9 above)
Comment (Manuthor) [reply crates/fpe/src/ffi/integer.rs]: done

File: crates/fpe/src/wasm_bindgen/float.rs (hunk 12 above)
Comment (Manuthor) [reply crates/fpe/src/wasm_bindgen/float.rs]: indeed

File: crates/fpe/src/wasm_bindgen/integer.rs (hunk 13 above)
Comment (Manuthor) [reply crates/fpe/src/wasm_bindgen/integer.rs]: done

**Pull request merged**